src/core/Akka/Serialization/NewtonSoftJsonSerializer.cs

```diff
@@ -107,40 +110,32 @@ public override byte[] ToBinary(object obj)
         public override object FromBinary(byte[] bytes, Type type)
         {
             string data = Encoding.Default.GetString(bytes);
-
+            JsonConvert.DefaultSettings =  () => _settings;
             object res = JsonConvert.DeserializeObject(data, _settings);
-            return TranslateSurrogate(res, system, type);
+            return TranslateSurrogate(res, this, type);
         }
 
-        private static object TranslateSurrogate(object deserializedValue,ActorSystem system,Type type)
+        private static object TranslateSurrogate(object deserializedValue,NewtonSoftJsonSerializer parent,Type type)
         {
             var j = deserializedValue as JObject;
             if (j != null)
             {
+                //The JObject represents a special akka.net wrapper for primitives (int,float,decimal) to preserve correct type when deserializing
                 if (j["$"] != null)
                 {
                     var value = j["$"].Value<string>();
                     return GetValue(value);
                 }
 
-                if (j["Case"] != null)
-                {
-                    var caseTypeName = j["Case"].Value<string>();
-                    var caseType = type.GetNestedType(caseTypeName);
-                    var ctor = caseType.GetConstructors(BindingFlags.NonPublic | BindingFlags.Instance)[0];
-                    var paramTypes = ctor.GetParameters().Select(p => p.ParameterType).ToArray();
-                    var values =
-                        j["Fields"].ToObject<object[]>()
-                            .Select((o, i) => TranslateSurrogate(o, system, paramTypes[i]))
-                            .ToArray();
-                    var res = ctor.Invoke(values.ToArray());
-                    return res;
-                }
+                //The JObject is not of our concern, let Json.NET deserialize it.
+                return j.ToObject(type, parent._serializer);
```

Comment:
This does the magic, if we get a JObject, we can pass in the same settings as we use for other serialzation, and things turns out correct :)